### PR TITLE
feat(ts/detours): add context and input to detour machine

### PR DESCRIPTION
**Pull request by firestack**
To be able to control the route and variant, the state machine needs to be the source of truth for the active route and pattern. This moves the values into the state machine and updates callers to send the route and pattern.

Asana Ticket: https://app.asana.com/0/0/1207481283409966/f

### CODE REVIEW

File: assets/src/components/detours/diversionPage.tsx [hunk 1]
```diff
@@ -294,3 +300,18 @@ DiversionPagePanel.Body = DiversionPagePanelBody
 DiversionPage.Panel = DiversionPagePanel
 
 export const Panel = DiversionPagePanel
+
+function descriptors(route: Route, routePattern: RoutePattern) {
```
Comment (firestack) [inline assets/src/components/detours/diversionPage.tsx:304]: todo: needs a better name

File: assets/src/components/detours/diversionPage.tsx [hunk 2]
```diff
@@ -137,7 +143,7 @@ export const DiversionPage = ({
           )}
           {routingError?.type === "unknown" && <RoutingErrorAlert />}
           <DetourMap
-            originalShape={originalRoute.shape.points}
+            originalShape={shape?.points ?? []}
```
Comment (firestack) [inline assets/src/components/detours/diversionPage.tsx:146]: note: is this a breaking change? Not sure what to do here since we can't derive shapes from the route pattern and expect that they're always there.

File: assets/src/components/detours/diversionPage.tsx (hunk 1 above)
Comment (firestack) [reply assets/src/components/detours/diversionPage.tsx:304]: also probably shouldn't be `function`

File: assets/src/hooks/useDetour.ts [hunk 3]
```diff
@@ -22,8 +22,12 @@ const useDetourDirections = (shapePoints: ShapePoint[]) =>
     }, [shapePoints]),
   })
 
-export const useDetour = ({ routePatternId, shape }: OriginalRoute) => {
-  const [snapshot, send] = useMachine(createDetourMachine)
+export const useDetour = (input: OriginalRoute) => {
```
Comment (firestack) [inline assets/src/hooks/useDetour.ts:25]: note:suggestion: I'd almost like to remove `OriginalRoute` here altogether and use, haven't come up with any downsides yet.
```ts
InputFrom<
  ActorLogicFrom<typeof createDetourMachine>
>
```

File: assets/src/models/createDetourMachine.ts [hunk 4]
```diff
@@ -1,11 +1,25 @@
...
+    input: {
+      // Caller has target route pattern
+      route: Route
+      routePattern: RoutePattern
+    }
     events: { type: "detour.edit.done" } | { type: "detour.edit.resume" }
   },
 }).createMachine({
   id: "Detours Machine",
+  context: ({ input }) => ({
```
Comment (firestack) [inline assets/src/models/createDetourMachine.ts:20]: note:nit: could be simplified

File: assets/tests/components/detours/diversionPage.test.tsx [hunk 5]
```diff
@@ -35,31 +33,23 @@ import {
...
+import { originalRouteFactory } from "../../factories/originalRouteFactory"
+import { DeepPartial } from "fishery"
 
 const DiversionPage = (
   props: Omit<
     Partial<ComponentProps<typeof DiversionPageDefault>>,
     "originalRoute"
   > & {
-    originalRoute?: Partial<
+    originalRoute?: DeepPartial<
```
Comment (firestack) [inline assets/tests/components/detours/diversionPage.test.tsx:44]: note: this could be it's own PR if desired.

File: assets/tests/components/detours/diversionPage.test.tsx [hunk 6]
```diff
@@ -1039,13 +1029,16 @@ describe("DiversionPage", () => {
...
+          route: {
+            name: routeName,
+          },
+          routePattern: {
+            headsign: routeDescription,
+            name: routeOrigin,
+            shape: {
+              points: [connectionPoint],
+            },
+          },
```
Comment (firestack) [inline assets/tests/components/detours/diversionPage.test.tsx:1041]: note: unsure about this due to how encapsulation is broken here.

Comment (joshlarson) [reply assets/tests/components/detours/diversionPage.test.tsx:1041]: IMO this is fine. The fact that `routedDescription`, for instance, comes from the `headsign` of the `routePattern` is a meaningful statement about how `<DiversionPage />` works, so it's not wrong to have the test be explicit about that here.

File: assets/src/hooks/useDetour.ts (hunk 3 above)
Comment (firestack) [reply assets/src/hooks/useDetour.ts:25]: This gives us the option to add more to `input` without having to change the `useDetour` interface to use it. going to add this to this PR.

File: assets/src/components/dummyDetourPage.tsx [hunk 7]
```diff
@@ -18,16 +18,11 @@ export const DummyDetourPage = () => {
...
-      {routePattern && routePattern.shape && (
+      {route && routePattern && routePattern.shape && (
         <DiversionPage
           originalRoute={{
-            shape: routePattern.shape,
-            routeName: routePattern.routeId,
-            routeDescription: routePattern.headsign || "?",
-            routeOrigin: routePattern.name,
-            routeDirection:
-              route?.directionNames[routePattern.directionId] || "?",
```
Comment (joshlarson) [inline assets/src/components/dummyDetourPage.tsx:29]: Lol I forgot that I put that in there as `"?"` - I would have been interested to see the bug report when a dispatcher saw `Inbound`/`Outbound`/`?`

File: assets/src/components/detours/diversionPage.tsx (hunk 1 above)
Comment (firestack) [reply assets/src/components/detours/diversionPage.tsx:304]: > Out of curiosity, how come you wanted this to be extracted as a function, rather than just inlined directly to `DiversionPage` itself?

It was just real messy and I was getting annoyed with how much noise it was adding

Comment (firestack) [reply assets/src/components/detours/diversionPage.tsx:304]: > And yeah, I _think_ our conventions are that this should be `const descriptors = (...) => {`?

Agreed, though I'm torn because the only reason this can be at the bottom of the file (which is where _I_ would prefer it) is due to hoisting. So converting this to `const` would mean it needs to live at the top of the file, before it's first invocation. So maybe we reconsider this practice at some point but I'll change it for now.

Given it'd have to be the first invocation, do you think in-lining it would be better?

Comment (joshlarson) [reply assets/src/components/detours/diversionPage.tsx:304]: Huhhhhhhh - I forgot about that. I think per what our team has been doing so far, it should be a `const` and live at the top of the file (I'm so used to that in our React code that it doesn't really bother me, but now that you mention it, it is weird that we like... sometimes have the "main component" at the top, and sometimes have it at the bottom). Happy to change our rules on that, but we should make an intentional choice to do that, and then follow the new rules.

No objection to in-lining, but no pressure to do so either.

Comment (firestack) [reply assets/src/components/detours/diversionPage.tsx]: @joshlarson what do you think of https://github.com/mbta/skate/compare/4829f3842bf9b7304a55362b17f8c5aa2aa74f80..e180162268a170207ddc2513135b1857d91ef998 ?

Comment (joshlarson) [reply assets/src/components/detours/diversionPage.tsx]: Probably fine, but since it's above anyway, does it need to be capitalized?

Comment (firestack) [reply assets/src/components/detours/diversionPage.tsx]: oh whoops, didn't mean to do that

Comment (firestack) [reply assets/src/components/detours/diversionPage.tsx]: https://github.com/mbta/skate/compare/b2625564eee19238849772cd11042fdccdff4e81..dca6719b349e19cc09ed41d051d104eed913fbbc

**Review by joshlarson — APPROVED** [review]
